Fix bt_home_hub_5 device tracker

ahobsonsayers opened the pull request:
## Description:
Updated BT Home Hub 5 device tracker component to get it working again. The old parsing method of the DNS table has been broken for a while causing the component to fail to get connected devices. A new parsing method has been implemened and fixes all previous issues. Home Assistant now succeeds in getting connected devices

No Other Changes.

ahobsonsayers committed:
Fix bt_home_hub_5 device tracker

Updated BT Home Hub 5 device tracker component to get it working again. The old parsing method of the DNS table has been broken for a while causing the component to fail to get connected devices. A new parsing method has been implemened and fixes all previous issues.

homeassistant commented:
Hi @ahobsonsayers, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#15096).

Once you do that we will be able to review and accept this pull request.

Thanks!

MartinHjelmare reviewed (changes requested):
There's too much device specific code going on in this platform, also before this PR. That code should be extracted to a standalone library published on pypi.

ahobsonsayers commented:
If you are referring to the HTMLTableParser class  (I assume you are as that is the only major section of code ive added) I  don't think I can extract it to a standalone library on PyPI as the  code is not mine. The repository containing the code can be found here: 

[html-table-parser-python3](https://github.com/schmijos/html-table-parser-python3)  

As  stated in that repository the author suggests just copying the class  located in parse.py into your own code. That is exactly what ive done to use it. The original author of the code has not added the library to  PyPI, and since I am not the owner, I don't belive the licence allows me  to add it to PyPI myself.  

That being said, I am not a lawyer so I don't know the ins and outs of the licensee.  Please advise on the best next steps to take as im not sure what now needs to be done to close this PR.

ahobsonsayers committed:
Moved part of code to a published PyPi library

ahobsonsayers commented:
Hi Martin, I have now moved the section of code you mentioned to a PyPi library as requested. I hope this satisfies the requirements and means that this PR can now be closed. Many Thanks!

ahobsonsayers committed:
Fixed Violations

ahobsonsayers committed:
Fixed bugs in device tracker

MartinHjelmare commented:
We need a standalone library that is specific for the device home assistant is interfacing with. Both the requests and data parsing should be moved to this library, ie both `_get_homehub_data` and `_parse_homehub_response`.

We should be able to instantiate a client class and retrieve relevant info (mac addresses etc) without knowing specific things about the device.

https://developers.home-assistant.io/docs/en/creating_platform_code_review.html#6-communication-with-devices-services

ahobsonsayers committed:
Moved API Specific Code to PyPi Repository

ahobsonsayers committed:
Merge branch 'dev' into bt-home-hub-5-fix

MartinHjelmare commented on homeassistant/components/device_tracker/bt_home_hub_5.py:
```diff
@@ -4,22 +4,20 @@
+
 _LOGGER = logging.getLogger(__name__)
-_MAC_REGEX = re.compile(r'(([0-9A-Fa-f]{1,2}\:){5}[0-9A-Fa-f]{1,2})')
+_MAC_REGEX = re.compile(r'(([0-9A-Fa-f]{1,2}:){5}[0-9A-Fa-f]{1,2})')
```
This isn't used anymore.

MartinHjelmare commented on homeassistant/components/device_tracker/bt_home_hub_5.py:
```diff
@@ -4,22 +4,20 @@
 For more details about this platform, please refer to the documentation at
 https://home-assistant.io/components/device_tracker.bt_home_hub_5/
 """
+
```
Please remove this blank line.

MartinHjelmare commented on homeassistant/components/device_tracker/bt_home_hub_5.py:
```diff
@@ -4,22 +4,20 @@
-from urllib.parse import unquote
-
-import requests
 import voluptuous as vol
```
Add a blank line between standard library and 3rd party imports.

MartinHjelmare commented on homeassistant/components/device_tracker/bt_home_hub_5.py:
```diff
@@ -37,44 +35,47 @@ class BTHomeHub5DeviceScanner(DeviceScanner):
+
         """Initialise the scanner."""
         _LOGGER.info("Initialising BT Home Hub 5")
         self.host = config.get(CONF_HOST, '192.168.1.254')
```
Move the default value to the config schema.

MartinHjelmare commented on homeassistant/components/device_tracker/bt_home_hub_5.py:
```diff
@@ -37,44 +35,47 @@ class BTHomeHub5DeviceScanner(DeviceScanner):
+        import bthomehub5_devicelist
+
         if not self.success_init:
             return False
```
Nothing is checking this return value.

MartinHjelmare commented on homeassistant/components/device_tracker/bt_home_hub_5.py:
```diff
@@ -83,45 +84,3 @@ def _update_info(self):
         self.last_results = data
 
         return True
```
Same as above.

MartinHjelmare commented on homeassistant/components/device_tracker/bt_home_hub_5.py:
```diff
@@ -37,44 +35,47 @@ class BTHomeHub5DeviceScanner(DeviceScanner):
+
+        import bthomehub5_devicelist
+
         if not self.success_init:
```
This looks like it can be removed.

MartinHjelmare reviewed (commented):
Nice!

Remember to run `script/gen_requirements.py` to update the requirements file correctly.

ahobsonsayers committed:
Updated to fit requested changes, removed test as it is no longer valid and updated requirement_all.txt

MartinHjelmare commented on homeassistant/components/device_tracker/bt_home_hub_5.py:
```diff
@@ -37,91 +36,47 @@ class BTHomeHub5DeviceScanner(DeviceScanner):
 
         if not data:
             _LOGGER.warning("Error scanning devices")
             return False
```
Just `return`.

ahobsonsayers committed:
Update to fit style requirements and remove redundant code

ahobsonsayers commented:
Hi Martin. Wanted to say thank you for the all the help. Im new to developing for home assistant as you probably gathered, so thanks for putting up with me and helping me get this component completed. The component has now passed all checks and I believe I have addressed all your comments so hopefully it should be good to go. Let me know if any other changes are needed.

MartinHjelmare commented on homeassistant/components/device_tracker/bt_home_hub_5.py:
```diff
@@ -65,10 +64,8 @@ def get_device_name(self, device):
-        """
-
+        """Ensure the information from the BT Home Hub 5 is up to date."""
+        # Return boolean if scanning successful.
```
We don't return a boolean anymore.

MartinHjelmare commented:
Looks good. Just fix the last comment and then we can merge.

ahobsonsayers committed:
Removed Unnecessary Comment

MartinHjelmare reviewed (approved):
Great!

MartinHjelmare commented:
Since we changed the config schema to make host optional, we should update the docs.

https://www.home-assistant.io/components/device_tracker.bt_home_hub_5/

Please link to a docs PR. After that we merge.

ahobsonsayers commented:
Updated the docs. Heres the PR:
https://github.com/home-assistant/home-assistant.io/pull/5964
